Data races in ingester

I've run the following and found some races:

```
❯ go test -mod=vendor -race  ./pkg/ingester
==================
WARNING: DATA RACE
Read at 0x00c00020c7f0 by goroutine 103:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bounds()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:458 +0x5e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepStream()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:174 +0x26e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepInstance()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:157 +0x16b
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepUsers()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:148 +0x90
  github.com/grafana/loki/pkg/ingester.(*Ingester).loop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:238 +0x112

Previous write at 0x00c00020c7f0 by goroutine 101:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).cut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:437 +0x378
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:256 +0x12a2
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Goroutine 103 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:187 +0x366
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 101 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8
==================
==================
WARNING: DATA RACE
Read at 0x00c0000c8020 by goroutine 103:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bounds()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:459 +0x3ba
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepStream()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:174 +0x26e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepInstance()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:157 +0x16b
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepUsers()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:148 +0x90
  github.com/grafana/loki/pkg/ingester.(*Ingester).loop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:238 +0x112

Previous write at 0x00c0000c8020 by goroutine 101:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).cut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:437 +0x337
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:256 +0x12a2
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Goroutine 103 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:187 +0x366
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 101 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8
==================
==================
WARNING: DATA RACE
Read at 0x00c0000c8028 by goroutine 103:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bounds()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:460 +0x43c
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepStream()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:174 +0x26e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepInstance()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:157 +0x16b
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepUsers()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:148 +0x90
  github.com/grafana/loki/pkg/ingester.(*Ingester).loop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:238 +0x112

Previous write at 0x00c0000c8028 by goroutine 101:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).cut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:437 +0x337
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:256 +0x12a2
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Goroutine 103 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:187 +0x366
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 101 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8
==================
==================
WARNING: DATA RACE
Read at 0x00c0007e71a0 by goroutine 103:
  github.com/grafana/loki/pkg/chunkenc.(*headBlock).isEmpty()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:98 +0xad
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bounds()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:463 +0x80
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepStream()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:174 +0x26e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepInstance()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:157 +0x16b
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepUsers()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:148 +0x90
  github.com/grafana/loki/pkg/ingester.(*Ingester).loop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:238 +0x112

Previous write at 0x00c0007e71a0 by goroutine 101:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).cut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:448 +0x456
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:256 +0x12a2
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Goroutine 103 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:187 +0x366
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 101 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8
==================
==================
WARNING: DATA RACE
Write at 0x00c00041c9b0 by goroutine 101:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).cut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:449 +0x49a
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:256 +0x12a2
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Previous read at 0x00c00041c9b0 by goroutine 103:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bounds()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:465 +0x32c
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepStream()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:174 +0x26e
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepInstance()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:157 +0x16b
  github.com/grafana/loki/pkg/ingester.(*Ingester).sweepUsers()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:148 +0x90
  github.com/grafana/loki/pkg/ingester.(*Ingester).loop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:238 +0x112

Goroutine 101 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 103 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:187 +0x366
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8
==================
--- FAIL: TestChunkFlushingIdle (20.61s)
    testing.go:906: race detected during execution of test
==================
WARNING: DATA RACE
Write at 0x00c00010e150 by goroutine 82:
  github.com/grafana/loki/pkg/ingester.(*Ingester).collectChunksToFlush()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:252 +0x676
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:216 +0xbf
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Previous read at 0x00c00010e150 by goroutine 83:
  github.com/grafana/loki/pkg/ingester.(*stream).Iterator()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/stream.go:261 +0x4e5
  github.com/grafana/loki/pkg/ingester.TestTransferOut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/transfer_test.go:90 +0x11d5
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb

Goroutine 82 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 83 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:80 +0x223
==================
==================
WARNING: DATA RACE
Write at 0x00c00010e120 by goroutine 82:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Bytes()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:283 +0xd6f
  github.com/grafana/loki/pkg/chunkenc.Facade.Marshal()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/facade.go:47 +0x4b
  github.com/grafana/loki/pkg/chunkenc.(*Facade).Marshal()
      <autogenerated>:1 +0xd5
  github.com/cortexproject/cortex/pkg/chunk.(*Chunk).Encode()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/chunk/chunk.go:220 +0x4dd
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushChunks()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:330 +0x6e9
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:224 +0x2c8
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Previous read at 0x00c00010e120 by goroutine 83:
  github.com/grafana/loki/pkg/chunkenc.(*MemChunk).Iterator()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/chunkenc/memchunk.go:481 +0x28d
  github.com/grafana/loki/pkg/ingester.(*stream).Iterator()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/stream.go:262 +0x621
  github.com/grafana/loki/pkg/ingester.TestTransferOut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/transfer_test.go:90 +0x11d5
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb

Goroutine 82 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 83 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:80 +0x223
==================
==================
WARNING: DATA RACE
Write at 0x00c00010e158 by goroutine 82:
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushUserSeries()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:231 +0x39f
  github.com/grafana/loki/pkg/ingester.(*Ingester).flushLoop()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/flush.go:196 +0x5cc

Previous read at 0x00c00010e158 by goroutine 83:
  github.com/grafana/loki/pkg/ingester.(*stream).Iterator()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/stream.go:261 +0x4e5
  github.com/grafana/loki/pkg/ingester.TestTransferOut()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/transfer_test.go:90 +0x11d5
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb

Goroutine 82 (running) created at:
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:171 +0xb3
  github.com/grafana/loki/pkg/ingester.(*Ingester).starting-fm()
      /Users/ctovena/go/src/github.com/grafana/loki/pkg/ingester/ingester.go:167 +0x55
  github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main()
      /Users/ctovena/go/src/github.com/grafana/loki/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x6f8

Goroutine 83 (running) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1042 +0x660
  testing.runTests.func1()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1284 +0xa6
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb
  testing.runTests()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1282 +0x527
  testing.(*M).Run()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:1199 +0x2ff
  main.main()
      _testmain.go:80 +0x223
==================
--- FAIL: TestTransferOut (0.31s)
    testing.go:906: race detected during execution of test
FAIL
FAIL	github.com/grafana/loki/pkg/ingester	89.175s
FAIL
```

I don't think they all are legitimate, and probably generated by the tests, but we should make sure this is the case and solve them all.

--- a/pkg/ingester/flush.go
+++ b/pkg/ingester/flush.go
@@ -3,6 +3,7 @@ package ingester
 import (
 	"fmt"
 	"net/http"
+	"sync"
 	"time"
 
 	"golang.org/x/net/context"
@@ -221,7 +222,7 @@ func (i *Ingester) flushUserSeries(userID string, fp model.Fingerprint, immediat
 	ctx := user.InjectOrgID(context.Background(), userID)
 	ctx, cancel := context.WithTimeout(ctx, i.cfg.FlushOpTimeout)
 	defer cancel()
-	err := i.flushChunks(ctx, fp, labels, chunks)
+	err := i.flushChunks(ctx, fp, labels, chunks, &instance.streamsMtx)
 	if err != nil {
 		return err
 	}
@@ -306,7 +307,7 @@ func (i *Ingester) removeFlushedChunks(instance *instance, stream *stream) {
 	}
 }
 
-func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelPairs labels.Labels, cs []*chunkDesc) error {
+func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelPairs labels.Labels, cs []*chunkDesc, streamsMtx *sync.RWMutex) error {
 	userID, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return err
@@ -327,7 +328,10 @@ func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, labelP
 		)
 
 		start := time.Now()
-		if err := c.Encode(); err != nil {
+		streamsMtx.Lock()
+		err := c.Encode()
+		streamsMtx.Unlock()
+		if err != nil {
 			return err
 		}
 		chunkEncodeTime.Observe(time.Since(start).Seconds())

--- a/pkg/ingester/transfer_test.go
+++ b/pkg/ingester/transfer_test.go
@@ -86,6 +86,7 @@ func TestTransferOut(t *testing.T) {
 		lines := []string{}
 
 		// Get all the lines back and make sure the blocks transferred successfully
+		ing2.instances["test"].streamsMtx.RLock()
 		for _, stream := range ing2.instances["test"].streams {
 			it, err := stream.Iterator(
 				context.TODO(),
@@ -103,7 +104,7 @@ func TestTransferOut(t *testing.T) {
 				lines = append(lines, entry.Line)
 			}
 		}
-
+		ing2.instances["test"].streamsMtx.RUnlock()
 		sort.Strings(lines)
 
 		assert.Equal(

